File: 125.valid-palindrome.py
#
# @lc app=leetcode id=125 lang=python3
#
# [125] Valid Palindrome
#
class Solution:
    def isPalindrome(self, s: str) -> bool:
        s=s.lower()
        if len(s)<=1:
            return False
        i=0,j=len(s)-1
        while i<j:
            while i<j and not s[i].isalnum():
                i=i+1
            while i<j and not s[j].isalnum():
                j=j+1
            if s[j]!=s[i]:
                return False
        return True


'''解法类似于c++解法 遇到数字和空格就跳过 s.lower 将所有字母都转为小写

File: 125.valid-palindrome.cpp
/*
 * @lc app=leetcode id=125 lang=cpp
 *
 * [125] Valid Palindrome
 */
class Solution {
public:
    bool isPalindrome(string s) {
        if(s.size()<=1) return true;
        int i=0,j=s.size()-1;
        while(i<j){
            while(i<j && !isalnum(s[i]))
            i++;
            while(i<j && !isalnum(s[j]))
            j--;
            if(tolower(s[i++])!=tolower(s[j--]))
            return false;

        }
        return true;
    }
};
/*双指针， 如果遇到非字母和数字则跳过，否则判断是否相等，在判断后需要注意将
i++，j--，以便while循环继续进行*/
